src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,223 @@
+def is_derangement {α : Type*} (f : perm α) : Prop := ∀ x : α, f x ≠ x
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
+
+/-- The permutation `f` has at most one fixed point, `a`. -/
+def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
+
+/-- The permutation `f` has exactly one fixed point, `a`. -/
+def exactly_one_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x ↔ x = a
+
+end definitions
+
+section simple_lemmas
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangements_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+subtype_equiv (perm_congr e) $ λ f, e.forall_congr $ λ x, by simp
+
+end simple_lemmas
+
+section fixed_points
+
+variables {α : Type*}
+
+lemma eofp_iff_opfp_and_eq (a : α) (f : perm α) :
+  exactly_one_fixed_point f a ↔ only_possible_fixed_point f a ∧ f a = a :=
+begin
+  split,
+  { intro h_eofp, exact ⟨λ x, (h_eofp x).mp, (h_eofp a).mpr rfl⟩ },
+  { rintros ⟨h_opfp, fa_eq_a⟩ x,
+    split,
+    { exact h_opfp x },
+    { rintro ⟨rfl⟩, exact fa_eq_a }}
+end
+
+lemma is_derangement_iff_opfp_and_ne (a : α) (f : perm α) :
+  is_derangement f ↔ only_possible_fixed_point f a ∧ f a ≠ a :=
+begin
+  split,
+  { intro h_derangement,
+    split,
+    { intros x x_fixed, exfalso, exact h_derangement x x_fixed },
+    { exact h_derangement a }},
+  { rintros ⟨h_opfp, fa_ne_a⟩ x x_fixed,
+    specialize h_opfp x x_fixed,
+    rw h_opfp at x_fixed,
+    exact fa_ne_a x_fixed }
+end
+
+/-- The set of permutations with `a` the only fixed point is equivalent to the set of derangements
+    on `{a}ᶜ`. -/
+def eofp_equiv_derangements_except_for [decidable_eq α] (a : α) :
+  {f : perm α | exactly_one_fixed_point f a} ≃ derangements ({a}ᶜ : set α) :=
+begin
+  let key_equiv : {f : perm α | f a = a} ≃ perm ({a}ᶜ : set α),
+  { transitivity {f : perm α // ∀ x : ({a} : set α), f x = (equiv.refl _) x},
+    { apply subtype_equiv_right, simp },
+    { exact equiv.set.compl (equiv.refl _) }},
+
+  transitivity {f : {f : perm α // f a = a} // only_possible_fixed_point f.val a},
+  { apply equiv.trans _ (subtype_subtype_equiv_subtype_exists _ _).symm,
+    apply subtype_equiv_right,
+    intro f,
+    simp_rw [exists_prop, set.mem_set_of_eq, eofp_iff_opfp_and_eq, and_comm] },
+  { refine subtype_equiv key_equiv _,
+    rintro ⟨f, _⟩,
+    simp [key_equiv, equiv.set.compl, derangements, is_derangement,
+          only_possible_fixed_point, not_imp_not] }
+end
+
+end fixed_points
+
+section option
+
+variables {α : Type*} [decidable_eq α]
+
+/-- The set of permutations `f` such that the preimage of `(a, f)` under
+    `perm.decompose_option` is a derangement. -/
+def remove_none.fiber (a : option α) : set (perm α) :=
+  {f : perm α | (a, f) ∈ equiv.perm.decompose_option '' derangements (option α)}
```

Comment:
This might be easier to work with as:
```suggestion
  {f : perm α | (equiv.perm.decompose_option.symm (a, f)).is_derangement}
```

Reply:
This setup makes it match `equiv.set_prod_equiv_sigma`, so I think it's easier to work with in the original form